File: python/unpack.py
import sys
import os
import json
import pprint
import copy
import collections
import _util
import decoders

def unpack_payload(payload, template):
    minors = payload.split("/")

    template = _util.Template(template)
    unpacked = template.template

    for li in xrange(0, len(template.columns)):
        packed_value = ""
        if li < len(minors): packed_value = minors[li] 

        rule = template.columns[li]

        decoder_name = rule.encoder.replace("-", "_")
        if not hasattr(decoders, decoder_name):
            raise Error("unknown decoding: ${rule.encoder}")
            continue

        decoder = getattr(decoders, decoder_name)
        unpacked_value = decoder(rule, packed_value)
        
        _util.set(unpacked, rule.path, unpacked_value)

    return unpacked

def unpack(packed, resolver_resolver):
    if not _util.isString(packed):
        raise Error("jsonxt.unpack: 'packed' not String")
    if not _util.isFunction(resolver_resolver):
        raise Error("jsonxt.unpack: 'resolver_resolver' not Function")

    majors = packed.split(":")

    if len(majors) < 5:
        raise Error("jsonxt.unpack: expected at least 5 parts")

    schema = _util.decode(majors[0].lower())
    resolver_name = _util.decode(majors[1].lower())
    type = _util.decode(majors[2].lower())
    version = _util.decode(majors[3].lower())

    if schema != "jxt":
        raise Error("jsonxt.unpack: unknown schema '${schema}'")

    templates = resolver_resolver(resolver_name, "templates.json")
    if _util.isString(templates):
        templates = json.loads(templates)
    if not _util.isPlainObject(templates):
        raise Error("jsonxt.unpack: 'templates' not Plain Object")

    type_version = "%(type)s:%(version)s" % {
        "type": type,
        "version": version,
    }

    template = templates[type_version]
    if not _util.isPlainObject(template):
        raise Error("jsonxt.unpack: 'templates['${type}:${version}']' not String")

    return unpack_payload(majors[4], template)
    
__dirname__ = os.path.dirname(os.path.abspath(__file__))
TESTS = os.path.join(__dirname__, "..", "test", "data")

with open(os.path.join(TESTS, "c4-1-1-packed.txt")) as fin:
    uri = fin.read()

def resolver(resolver_name, resolver_type):
    with open(os.path.join(TESTS, "covid-templates.json")) as fin:
        return fin.read()

result = unpack(uri, resolver)
print(json.dumps(result, indent=2))

File: javascript/lib/unpack.js
/**
 *  lib/unpack.js
 *
 *  David Janes
 *  Consensas
 *  2021-03-12
 *
 *  Copyright (2013-2021) Consensas
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

"use strict"

const _util = require("./_util")


/**
 *  Unpacks from $payloadStack an array of $arraySize elements of the type $template 
 */
const unpack_array = async (payloadStack, arraySize, template, templates) => {
    if (arraySize == null ||  arraySize == undefined  ||  arraySize === "") {
        return null;
    }

    let array = [];
    for (let index = 0; index < parseInt(arraySize); index++) {
        array.push(await unpack_template(payloadStack, template, templates));
    }
    return array;
} 

/**
 *  Unpacks from $payloadStack all fields of the type $template 
 */
const unpack_template = async (payloadStack, template, templates) => {
    const jsonxt = require("..")
    const unpacked = JSON.parse(JSON.stringify(template.template || {}))

    if (template && template.columns) {
        for (let li = 0; li < template.columns.length && payloadStack.length > 0; li++) {
            let payloadElement = payloadStack.shift();

            const rule = template.columns[li]

            if (rule.encoder === "array") {
                const arraySize = payloadElement;
                const nestedTemplate = templates[rule.encoder_param];
                if (!_util.isPlainObject(nestedTemplate)) {
                    throw new Error(`jsonxt.unpack: 'templates["${rule.encoder_param}"]' not String`)
                }

                const unpackedArray = await unpack_array(payloadStack, arraySize, nestedTemplate, templates);
                if (unpackedArray) {
                    _util.set(unpacked, rule.path, unpackedArray);
                }
            } else {
                const decoder = jsonxt.decoders[rule.encoder]
                if (!decoder) {
                    throw new Error(`unknown decoding: ${rule.encoder}`)
                }

                const unpacked_value = decoder(rule, payloadElement)
                if (_util.isUndefined(unpacked_value)) {
                    continue
                }

                _util.set(unpacked, rule.path, unpacked_value)
            }
        }
    }

    return unpacked
}

/**
 *  
 */
const remove_term_map = async (array) => {
    array.forEach(function(key, index) {
        if (key.charAt(0) === "*"){ 
            const mappedIndex = parseInt(_util.base32_to_integer(key.slice(1)));
            array[index] = array[mappedIndex];
        } 
    });
    return array;
}

/**
 *  Creates a stack of payload fields to unpack and starts the recursion 
 */
const unpack_payload = async (payload, template, templates) => {
    // Rebuilds sequences of null fields that were mapped into ' ' + index. 
    // Reverts * and $ as field separators: * -> /* and $ -> /$
    // TODO: This is UGLY. Need to find a better way to run over all these regex in order.  
    const dup = payload
        .replace(/ G/g, "///////////////////")
        .replace(/ F/g, "//////////////////")
        .replace(/ E/g, "/////////////////")
        .replace(/ D/g, "////////////////")
        .replace(/ C/g, "///////////////")
        .replace(/ B/g, "//////////////")
        .replace(/ A/g, "/////////////")
        .replace(/ 9/g, "////////////")
        .replace(/ 8/g, "///////////")
        .replace(/ 7/g, "//////////")
        .replace(/ 6/g, "/////////")
        .replace(/ 5/g, "////////")
        .replace(/ 4/g, "///////")
        .replace(/ 3/g, "//////")
        .replace(/ 2/g, "/////")
        .replace(/ 1/g, "////")
        .replace(/ 0/g, "///")
        .replace(/\*/g, "/*")
        .replace(/\$/g, "/$")

    // Splits string into fields. 
    const payloadStack = dup.split('/');
    // Removes the term map
    const expanded = await remove_term_map(payloadStack);
    // Standard JSONXT
    return await unpack_template(expanded, template, templates)
}

/**
 */
const unpack = async (packed, resolver_resolver) => {
    const jsonxt = require("..")

    if (!_util.isString(packed)) {
        throw new Error("jsonxt.unpack: 'packed' not String")
    }
    if (!_util.isFunction(resolver_resolver)) {
        throw new Error("jsonxt.unpack: 'resolver_resolver' not Function")
    }

    const majors = packed.split(":")

    if (majors.length < 5) {
        throw new Error(`jsonxt.unpack: expected at least 5 parts`)
    }

    const schema = _util.decode(majors[0].toLowerCase())
    const resolver_name = _util.decode(majors[1].toLowerCase())
    const type = _util.decode(majors[2].toLowerCase())
    const version = _util.decode(majors[3].toLowerCase())

    if (schema !== "jxt") {
        throw new Error(`jsonxt.unpack: unknown schema "${schema}"`)
    }

    let templates = await resolver_resolver(resolver_name, "templates.json")
    if (_util.isString(templates)) {
        templates = JSON.parse(templates)
    }
    if (!_util.isPlainObject(templates)) {
        throw new Error("jsonxt.unpack: 'templates' not Plain Object")
    }

    const type_version = `${type}:${version}`

    const template = templates[type_version]
    if (!_util.isPlainObject(template)) {
        throw new Error(`jsonxt.unpack: 'templates["${type}:${version}"]' not String`)
    }

    return await unpack_payload(majors[4], template, templates)
}

/**
 */
exports.unpack = unpack
exports.unpack.payload = unpack_payload
